design/open/hard-delete.md

```diff
@@ -0,0 +1,202 @@
+2. Object path is outside the repo namespace:
+   1. lakeFS will not handle its retention - therefore it will have to be deleted by the
+   user or by a dedicated GC process
+
+#### Delete Object
+
+1. Read object's metadata
+2. Reference counter exists:
+   1. if counter == 1
+      1. Hard-delete object
+   2. if counter > 1
+      1. Decrement counter 
+3. If reference counter doesn't exist:
+   1. Retention is not handled in lakeFS
+
+#### LinkPhysicalAddress
+
+Assume this API uses a physical address in the repo namespace
+1. Read object's metadata
+2. Increment counter and write metadata
+3. Add staging entry
+
+#### StageObject
+
+If physical address in repo namespace - perform increment as described above, otherwise do not handle retention
+
+#### Reset Branch
+
+When resetting a branch we throw all the uncommitted branch data.
+We can leverage the new KV drop async functionality to also check and hard-delete objects as needed
+
+### Atomic reference counter
+
+For each storage adapter (AWS, GC, AZ), we will use a builtin logic to provide this functionality
+
+#### AWS
+
+AWS doesn't allow to update an object or its metadata after creation. In order to update the metadata a copy object should be performed with the new metadata values. The only 
+conditionals currently available are on the objects timestamp and ETag. Unfortunately ETags are modified only on objects content change and does not take into account changes in metadata.
+
+#### Azure
+
+Store the reference counter in the blob's metadata and use `set blob metadata` API with the supported conditional header on the ETag to perform an If-Match operation and update
+the reference counter
+
+
+#### Google Store
+
+Store the reference counter in the blob's metadata, use the `patch object` API with the `ifMetagenerationMatch` conditional.
+
+
+## 3. Tracking references in staging manager
+
+Staging manager will track physical addresses of staged objects and the staging tokens referencing them.
+We will introduce a new object to the database:
+
+      key    = <repo>/references/<physical_address>/<staging_token>  
+      value  = state enum (staged|deleted|comitted)
+
+On Upload object to a branch, we will add a key in the references path with the physical address and current staging token and 
+mark it as staged.   
+On Delete object from a staging area, we will update the reference key with value `deleted`  
+On Commit, we will update the reference key with value `comitted`
+
+A background job will be responsible for scanning the references prefix, and handling the references according to the state.
+
+
+### Background delete job (pseudo-code)
+
+1. Scan `reference` prefix (can use after prefix for load balancing)
+2. For each `physical_address` read all entries
+      1. If found state == 'committed' in any entry
+         1. Delete all keys for `phisycal_address`, by order of state: deleted -> staged -> committed
```

Comment:
This seems worrying:
1. I commit an object at physical path `abcd` during staging stg1.  So I have `committed` at `references/abcd/stg1`.
1. Concurrently:
   * Background delete runs and deletes this entry (the intent is to keep `abcd`, it was committed).
   * A copy operation stages `abcd` into staging stg2
   After both of these, I can have `staged` at `references/abcd/stg2` and nothing (no key!) at `references/abcd/stg1`.
1. Delete this object.  Now `references/abcd/stg2` is `deleted`.
1. The next background delete deletes `abcd`, despite it being referenced by a commit.

Reply:
> Background delete runs and deletes this entry (the intent is to keep abcd, it was committed).
This will not happen since we will not delete references which have at least one entry with state `committed`

Additionally, something that was discussed and was omitted from the design (will add it), is that the background delete operation is time based. i.e. a candidate for delete is a physical address which all of its references are in state `deleted` and the last modified timestamp is bigger than 'x' (for example a day).
This ensures us that this physical address is no longer being in use or have a chance to be referenced again.